grammar/grammar.py

```diff
@@ -42,8 +47,13 @@
 # TODO consider a way to functionalize scope parses for reusability (don't break trampoline rule)
 
 class Grammar:
-    def __init__(self, rules:dict):
+    def __init__(self, rules:dict, function_library:dict={}):
```

Comment:
Try to avoid setting defaults to iterables if at all possible. Python will use that the next time it needs the default value for an instance of this class.

Instead set it to None and then if its None, set it to the default iterable you want.